### PR TITLE
Add test cases for antispam rules, fix bug in attachments rule & test

### LINKED ISSUE #595: Write unit tests for `bot/rules/burst.py`
Write unit tests for [`bot/rules/burst.py`](../blob/master/bot/rules/burst.py).

## Implementation details
Please make sure to read the general information in the [meta issue](553) and the [testing README](../blob/master/tests/README.md). We are aiming for a 100% [branch coverage](https://coverage.readthedocs.io/en/stable/branch.html) for this file, but if you think that is not possible, please discuss that in this issue.

## Additional information
If you want to work on this issue, **please make sure that you get assigned to it** by one of the core devs before starting to work on it. We would like to prevent the situation that multiple people are working on the same issue. To get assigned, leave a comment showing your interesting in tackling this issue.

### LINKED ISSUE #596: Write unit tests for `bot/rules/burst_shared.py`
Write unit tests for [`bot/rules/burst_shared.py`](../blob/master/bot/rules/burst_shared.py).

## Implementation details
Please make sure to read the general information in the [meta issue](553) and the [testing README](../blob/master/tests/README.md). We are aiming for a 100% [branch coverage](https://coverage.readthedocs.io/en/stable/branch.html) for this file, but if you think that is not possible, please discuss that in this issue.

## Additional information
If you want to work on this issue, **please make sure that you get assigned to it** by one of the core devs before starting to work on it. We would like to prevent the situation that multiple people are working on the same issue. To get assigned, leave a comment showing your interesting in tackling this issue.

### LINKED ISSUE #597: Write unit tests for `bot/rules/chars.py`
Write unit tests for [`bot/rules/chars.py`](../blob/master/bot/rules/chars.py).

## Implementation details
Please make sure to read the general information in the [meta issue](553) and the [testing README](../blob/master/tests/README.md). We are aiming for a 100% [branch coverage](https://coverage.readthedocs.io/en/stable/branch.html) for this file, but if you think that is not possible, please discuss that in this issue.

## Additional information
If you want to work on this issue, **please make sure that you get assigned to it** by one of the core devs before starting to work on it. We would like to prevent the situation that multiple people are working on the same issue. To get assigned, leave a comment showing your interesting in tackling this issue.

### LINKED ISSUE #598: Write unit tests for `bot/rules/discord_emojis.py`
Write unit tests for [`bot/rules/discord_emojis.py`](../blob/master/bot/rules/discord_emojis.py).

## Implementation details
Please make sure to read the general information in the [meta issue](553) and the [testing README](../blob/master/tests/README.md). We are aiming for a 100% [branch coverage](https://coverage.readthedocs.io/en/stable/branch.html) for this file, but if you think that is not possible, please discuss that in this issue.

## Additional information
If you want to work on this issue, **please make sure that you get assigned to it** by one of the core devs before starting to work on it. We would like to prevent the situation that multiple people are working on the same issue. To get assigned, leave a comment showing your interesting in tackling this issue.

### LINKED ISSUE #599: Write unit tests for `bot/rules/duplicates.py`
Write unit tests for [`bot/rules/duplicates.py`](../blob/master/bot/rules/duplicates.py).

## Implementation details
Please make sure to read the general information in the [meta issue](553) and the [testing README](../blob/master/tests/README.md). We are aiming for a 100% [branch coverage](https://coverage.readthedocs.io/en/stable/branch.html) for this file, but if you think that is not possible, please discuss that in this issue.

## Additional information
If you want to work on this issue, **please make sure that you get assigned to it** by one of the core devs before starting to work on it. We would like to prevent the situation that multiple people are working on the same issue. To get assigned, leave a comment showing your interesting in tackling this issue.

### LINKED ISSUE #602: Write unit tests for `bot/rules/newlines.py`
Write unit tests for [`bot/rules/newlines.py`](../blob/master/bot/rules/newlines.py).

## Implementation details
Please make sure to read the general information in the [meta issue](553) and the [testing README](../blob/master/tests/README.md). We are aiming for a 100% [branch coverage](https://coverage.readthedocs.io/en/stable/branch.html) for this file, but if you think that is not possible, please discuss that in this issue.

## Additional information
If you want to work on this issue, **please make sure that you get assigned to it** by one of the core devs before starting to work on it. We would like to prevent the situation that multiple people are working on the same issue. To get assigned, leave a comment showing your interesting in tackling this issue.

### LINKED ISSUE #603: Write unit tests for `bot/rules/role_mentions.py`
Write unit tests for [`bot/rules/role_mentions.py`](../blob/master/bot/rules/role_mentions.py).

## Implementation details
Please make sure to read the general information in the [meta issue](553) and the [testing README](../blob/master/tests/README.md). We are aiming for a 100% [branch coverage](https://coverage.readthedocs.io/en/stable/branch.html) for this file, but if you think that is not possible, please discuss that in this issue.

## Additional information
If you want to work on this issue, **please make sure that you get assigned to it** by one of the core devs before starting to work on it. We would like to prevent the situation that multiple people are working on the same issue. To get assigned, leave a comment showing your interesting in tackling this issue.

### PR DESCRIPTION
**Pull request by kwzrd**
This PR introduces a series of unit test cases for the bot's antispam rules. This addresses the following issues.

Closes #595 via 1f17d0e.
Closes #596 via b0713be.
Closes #597 via f15f318.
Closes #598 via 4bfe30d.
Closes #603 via f32ffaf.

I also sneak in a quick bugfix in 4199b12: it appears that the already existing rule for attachments contains a slight bug where in the printout, the configured `'max'` key is looked up instead of `'interval'`. The same commit also fixes the key in the rule's unit test, which was incorrectly expecting the wrong key.

Finally, aedbd6f is a refactor of the name `msg` into `make_msg` which prevents a harmless, but unnecessary name collision in the `msg for msg in ...` generator expressions down the module.

Edit:
Closes #599 via 54e8c98.
Closes #602 via cae7f25.

### CODE REVIEW

**Comment by kwzrd** [comment]
The committed code is considered ready for review. I'm currently awaiting a response from a user in #599 - if they're no longer interested, I'd like to include that issue in this PR's scope as well.

File: tests/bot/rules/test_burst.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import unittest
+
+from bot.rules import burst
+from tests.helpers import MockMessage, async_test
+
+
+def make_msg(author: str) -> MockMessage:
+    """Init a MockMessage instance with author set to `author`.
```
[the remark this answers is not in the document]
Comment (kwzrd) [reply tests/bot/rules/test_burst.py]: Thanks, addressed in 8456331.

**Comment by kwzrd** [comment]
The test cases are similar, yes - the rules are too. What do you propose can be done to make this better? Would it make sense to implement some kind of an "executor" on the side, and then just feed it the test cases from each module?

**Comment by MarkKoz** [comment]
Two helper functions: one for "allows" tests and one for "disallows". They should be passed the `apply` function and the subtest arguments. The helpers will do the loop, calling of `apply`, and the assertions.

As I mentioned, some of the subtest arguments follow a similar pattern across test suites. To facilitate this, a function could be defined which will take a `make_msg` function and return subtest arguments. This function can shared among some test suites to generate subtest arguments while other suites with more divergent subtest can still use their own subtest arguments. My only concern is that such function adds a layer of indirection which has the consequence of making unclear what the actual subtest values are.

I believe such a design will reduce redundancy but still remain flexible should test suites need to diverge in code in the future.

**Comment by kwzrd** [comment]
93d19f3 adds an ABC (9758e32 actually adds the `metaclass=ABCMeta`), b89f9c5 then adjusts the existing rules to use the ABC. I decide to make the `DisallowedCase` (previously `Case`) more generic by having an `n_violations` attr instead of a rule-specific one. This way, the `subTest` kwargs can be generic as well.

Let me know how this looks, and thanks for the feedback.

File: tests/bot/rules/__init__.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    Tests for specific rules should inherit from `RuleTest` and implement
+    `relevant_messages` and `get_report`. Each instance should also set the
+    `apply` and `config` attributes as necessary.
+
+    The execution of test cases can then be delegated to the `run_allowed`
+    and `run_disallowed` methods.
+    """
+
+    apply: Callable  # The tested rule's apply function
+    config: Dict[str, int]
```
Comment (MarkKoz) [inline tests/bot/rules/__init__.py]: These _could_ be enforced

```py
    @property
    @abc.abstractmethod
    def apply(self) -> Callable:
        """Docstring"""
        raise NotImplementedError
```

Comment (kwzrd) [reply tests/bot/rules/__init__.py]: I briefly considered this, but eventually decided not to. Should the child unittest not provide an `apply` attr, the tests will fail with:

```python
AttributeError: 'MyRuleTest' object has no attribute 'apply'
```

I believe that's descriptive enough. I'll mark this as resolved then if that's ok.

**Comment by kwzrd** [comment]
I've added tests for `duplicates` (54e8c98) and `newlines` (cae7f25).

As the latter is more complex than the other rules, yielding two distinct error reports, I decided to implement two subclasses of the `RuleTest` ABC for it. Each checks one type of violation.

If the given messages violate both **total count** and **max consecutive**, the rule will fail on the **total** as that is checked first. Since I consider this an implementation detail and not "desired functionality", I do not test it, and each disallowed test case only violates the rule one way or the other, so that the tests still pass should the two checks switch order.

This PR now provides 100% coverage for the entire `rules` package - I've updated the PR description to mention the two extra issues it closes.

**Pull request merged**